lightstep/sdk/metric/aggregator/histogram/exponential.go

```diff
@@ -0,0 +1,468 @@
+		growTo = s.maxSize
+	}
+	part := growTo - bias
+	switch counts := b.backing.(type) {
+	case []uint8:
+		tmp := make([]uint8, growTo)
+		copy(tmp[part:], counts[diff:])
+		copy(tmp[0:diff], counts[0:diff])
+		b.backing = tmp
+	case []uint16:
+		tmp := make([]uint16, growTo)
+		copy(tmp[part:], counts[diff:])
+		copy(tmp[0:diff], counts[0:diff])
+		b.backing = tmp
+	case []uint32:
+		tmp := make([]uint32, growTo)
+		copy(tmp[part:], counts[diff:])
+		copy(tmp[0:diff], counts[0:diff])
+		b.backing = tmp
+	case []uint64:
+		tmp := make([]uint64, growTo)
+		copy(tmp[part:], counts[diff:])
+		copy(tmp[0:diff], counts[0:diff])
+		b.backing = tmp
+	default:
+		panic("grow() with size() == 0")
+	}
+}
+
+// downscale first rotates, then collapses 2**`by`-to-1 buckets
+func (b *buckets) downscale(by int32) {
+	b.rotate()
+
+	size := 1 + b.indexEnd - b.indexStart
+	each := int64(1) << by
+	inpos := int32(0)
+	outpos := int32(0)
+
+	for pos := b.indexStart; pos <= b.indexEnd; {
+		mod := int64(pos) % each
+		if mod < 0 {
+			mod += each
+		}
+		for i := mod; i < each && inpos < size; i++ {
+			b.relocateBucket(outpos, inpos)
+			inpos++
+			pos++
+		}
+		outpos++
+	}
+
+	b.indexStart >>= by
+	b.indexEnd >>= by
+	b.indexBase = b.indexStart
+}
+
+// rotate shifts the backing array contents so that indexStart ==
+// indexBase to simplify the downscale logic.
+func (b *buckets) rotate() {
+	bias := b.indexBase - b.indexStart
+
+	if bias == 0 {
+		return
+	}
+
+	// Rotate the array so that indexBase == indexStart
+	b.indexBase = b.indexStart
+
+	b.reverse(0, b.size())
+	b.reverse(0, bias)
+	b.reverse(bias, b.size())
+}
+
+func (b *buckets) reverse(from, limit int32) {
+	num := ((from + limit) / 2) - from
+	switch counts := b.backing.(type) {
+	case []uint8:
+		for i := int32(0); i < num; i++ {
+			counts[from+i], counts[limit-i-1] = counts[limit-i-1], counts[from+i]
+		}
+	case []uint16:
+		for i := int32(0); i < num; i++ {
+			counts[from+i], counts[limit-i-1] = counts[limit-i-1], counts[from+i]
+		}
+	case []uint32:
+		for i := int32(0); i < num; i++ {
+			counts[from+i], counts[limit-i-1] = counts[limit-i-1], counts[from+i]
+		}
+	case []uint64:
+		for i := int32(0); i < num; i++ {
+			counts[from+i], counts[limit-i-1] = counts[limit-i-1], counts[from+i]
+		}
+	}
+}
+
+// relocateBucket adds the count in counts[src] to counts[dest] and
+// resets count[src] to zero.
+func (b *buckets) relocateBucket(dest, src int32) {
+	if dest == src {
+		return
+	}
+	switch counts := b.backing.(type) {
+	case []uint8:
+		tmp := counts[src]
+		counts[src] = 0
+		b.incrementBucket(dest, uint64(tmp))
+	case []uint16:
+		tmp := counts[src]
+		counts[src] = 0
+		b.incrementBucket(dest, uint64(tmp))
+	case []uint32:
+		tmp := counts[src]
+		counts[src] = 0
+		b.incrementBucket(dest, uint64(tmp))
+	case []uint64:
+		tmp := counts[src]
+		counts[src] = 0
+		b.incrementBucket(dest, uint64(tmp))
+	}
+}
+
+// incrementBucket increments the backing array index by `incr`.
```

Comment:
While I have come to appreciate the elegance of this approach, it took a few minutes of head-scratching to understand at all, and several more minutes of tinkering to see _why_ it's elegant. I would love to see a bit of elucidation in the docstring here.